Safety/v1

cryogenian opened the pull request:
This is very WIP, but the main things are implemented
+ All effects are terminal and datatype are concrete. 
+ `Decoder` and `Encoder` are everywhere.

wemrysi commented on core/src/main/scala/mongo4cats/client/ClientSession.scala:
```diff
@@ -16,72 +16,34 @@
 import mongo4cats.helpers._
 
-import scala.util.Try
+final case class ClientSession(session: JClientSession) {
```
I think an interface with an impl in the companion would be preferable (and more flexible) to a final class (also doesn't need to be a `case class`).

wemrysi reviewed (commented):
First off, thanks for tackling this!

Some general comments

1. After seeing example usage, I don't think pushing the effect type to the methods is an improvement. The usability/ergonomics seem quite poor since we have to ascribe the effect type and often the output type. I suggest we revert to making the traits parametric in the effect type, and maybe instead we push the typeclass evidence to the methods.
2. Especially as this is a library, I think we want to avoid `final class`es as the only representation of an inteface. A more flexible pattern is to keep the interface as a `trait` and have the private impl as a `final class`, for example. I noticed a handful of instances where we replaced a `trait` with a `final class`. We definitely don't want to be exposing public case classes as they make it tough to preserve binary compatibility.
3. We'd really like to avoid using `Document` at all, this brings back all the unsafety we're trying to avoid. I think we should prefer `BsonDocument` instead.

wemrysi commented:
Another issue with pushing all the effects to the leaves is it pushes `Async` constraints further down, which is undesirable imo. We can sort of mitigate this in applications by wrapping the mongo classes with specialized types, but I still think it is a net loss.

cryogenian commented on core/src/main/scala/mongo4cats/bson/Decoder.scala:
```diff
@@ -0,0 +1,78 @@
+  implicit val bsonDecoder: Decoder[BsonValue] = new Decoder[BsonValue] {
+    def apply(b: BsonValue) = Right(b)
+  }
+  implicit val document: Decoder[Document] = new Decoder[Document] {
```
Yep, it's used in examples and in at least one place in services, what do you mean that it's relies on codec machinery? I mean `parse` could use codecs as well as `toJson` but why should we care about it, it's hidden.

wemrysi replied:
`Document` exposes lots of [unsafe methods](https://mongodb.github.io/mongo-java-driver/4.2/apidocs/bson/org/bson/Document.html) that rely on the untyped java codec machinery. I don't think we want to encourage it's use and replace it with `BsonDocument` instead.

cryogenian commented on core/src/main/scala/mongo4cats/bson/Decoder.scala:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+trait DocumentDecoder[A] extends Serializable { self =>
```
Not really, this is doable, but why do we need this?

cryogenian commented on core/src/main/scala/mongo4cats/bson/package.scala:
```diff
@@ -47,19 +44,15 @@ object bson {
-    implicit val codecProvider: MongoCodecProvider[Document] = new MongoCodecProvider[Document] {
-      override def get: CodecProvider = new DocumentCodecProvider()
-    }
+    def parse(json: String): Document = JDocument.parse(json)
```
Oh, right! I see `Document implements Map[String, Object]`, right! I'd switched to `BsonDocument` in services. I'll remove this thing :+1:

cryogenian commented:
I agree that `Document` is problematic, and definitely should be either removed or moved to something like `mongo4cats.unsafe`. 

But I disagree with the first and second things. 
1.1. There is no any common used effects between methods, i.e. it's not necessary to have a natural transformation to do `mapK`. I.e. if we don't have any of knots there and we don't need any natural transformation to map, why provide the same effect for all of them? 
1.2. It is necessary to have type ascription on find/aggregate/distinct, the data persisted in mongo is untyped, how else we could say what is it? 
2. I agree that `ClientSession` might be exposed as a trait, e.g. for mocking reasons. I'm not that sure about queries and operations. 
4. Right, `Async` are pushed down (or up, idk) to callsites, which implies generalizing things over `Async` instead of `Functor` or `Monad` in some places. But isn't that correct? I mean there couldn't be `MongoCollection.find.stream[F, A]` where `F` isn't `Async`, right? I mean that we have `Persistence[F, K, V]` and it's `F[_]` is absolutely generic, but how could we have `MongoPersistence[F, K, V]` where `F[_]` isn't `Async`?

wemrysi commented:
> 1.1. There is no any common used effects between methods, i.e. it's not necessary to have a natural transformation to do `mapK`. I.e. if we don't have any of knots there and we don't need any natural transformation to map, why provide the same effect for all of them?

You're correct, but I think the improved ergonomics are worth it and, on the contrary, how often does one use different effects in a single context?

> 1.2. It is necessary to have type ascription on find/aggregate/distinct, the data persisted in mongo is untyped, how else we could say what is it?

Agreed that we need the ascription for the output type, I was referring to having to ascribe the effect type also.

> 2. I agree that `ClientSession` might be exposed as a trait, e.g. for mocking reasons. I'm not that sure about queries and operations.

It isn't just mocking, more for encapsulation and to avoid overly constraining types (part of why this PR exists is that some types upstream were over constrained). What downside is there to making them traits? Maybe some additional code, but I don't think that is excessive for the benefits.

> 4. Right, `Async` are pushed down (or up, idk) to callsites, which implies generalizing things over `Async` instead of `Functor` or `Monad` in some places. But isn't that correct? I mean there couldn't be `MongoCollection.find.stream[F, A]` where `F` isn't `Async`, right? I mean that we have `Persistence[F, K, V]` and it's `F[_]` is absolutely generic, but how could we have `MongoPersistence[F, K, V]` where `F[_]` isn't `Async`?

It reduces local reasoning via the constraints, i.e.

```scala
def foo[F: Monad](mongo: MongoCollection[F]): F[Int]

def bar[F: Async](mongo: MongoCollection[F]): F[Int]
```

when considering `foo`, I can rely on the compiler to enforce that it does pure things (with the general caveat that we as programmers agree to also adhere to the constraints and not cheat), possibly involving mongo collection operations (which themselves may do `Async` things, but I can't know that or run them here). `bar` however can do literally anything.

cryogenian commented:
> You're correct, but I think the improved ergonomics are worth it and, on the contrary, how often does one use different effects in a single context?

Well, we do use it in different context quite often 😄 

> It isn't just mocking, more for encapsulation and to avoid overly constraining types (part of why this PR exists is that some types upstream were over constrained). What downside is there to making them traits? Maybe some additional code, but I don't think that is excessive for the benefits.

There was one problem with `mapK` over queries, I think I found a solution to it, but not sure exactly, need to try it out. 

> It reduces local reasoning via the constraints, i.e.

Right, I agree. 

I think that enclosing `F[_]` over `MongoCollection` might be useful, yes, just need to come up how to convert things w/o need of mapping :confused:

rintcius commented:
> There was one problem with mapK over queries, I think I found a solution to it, but not sure exactly, need to try it out.

Hmm, curious what the problem there is

cryogenian commented:
> Hmm, curious what the problem there is

@rintcius 

Amount of code mostly, current encoding allows to do everything as needed.

cryogenian commented:
This is ready for review. I've removed 3.1.0 support, since I'm not that familiar with it, I guess that it doesn't work because implicit resolution is different with 2.13.7. Shouldn't be a big deal to fix, but I need to port services to this. 

 Effects are enclosed in classes/traits again, implemented `mapK` and `asK[F[_]: Async]` everywhere (no `asK` for queries).

domaspoliakas commented on core/src/main/scala/mongo4cats/client/MongoClient.scala:
```diff
@@ -16,55 +16,75 @@
+
+    def getDatabase(name: String) = transform {
+      Async[F].delay {
+        MongoDatabase[F](client.getDatabase(name)).mapK(transform)
```
I prefer the `transform(Async[F].delay(client.getDatabase(name)).map(MongoDatabase(_).mapK(transform)))` form to be honest. I don't know how much of an effect this would have in practice, but I prefer the `.delay` to be as small as possible.

It would probably matter here, but just my 2 cents.

cryogenian replied:
Ok :+1:

domaspoliakas commented on core/src/main/scala/mongo4cats/collection/operations/Filter.scala:
```diff
@@ -18,656 +18,169 @@ package mongo4cats.collection.operations
+    Filter(Filters.eq(fieldName, null))
+
+  def eq[A: Encoder](fieldName: String, value: A) = {
+    println("&&&&&&")
```
👀

cryogenian replied:
Thanks for spotting this!

domaspoliakas reviewed (commented):
This looks like a huge amount of work, well done!

cryogenian commented on circe/src/main/scala/mongo4cats/circe.scala:
```diff
@@ -16,62 +16,41 @@
-import org.bson.{BsonReader, BsonType, BsonWriter, Document}
-
-import scala.reflect.ClassTag
+import cats.implicits._
```
:+1: 

Will add to next commit with `Encoder` rename

cryogenian commented on core/src/main/scala/mongo4cats/bson/Encoder.scala:
```diff
@@ -0,0 +1,108 @@
+
+import org.bson._
+
+trait Encoder[A] extends Serializable { self =>
```
:+1:

cryogenian commented on core/src/main/scala/mongo4cats/client/ClientSession.scala:
```diff
@@ -16,72 +16,63 @@
-  private[client] def make[F[_]: Async](session: JClientSession): F[ClientSession[F]] =
-    Monad[F].pure(new LiveClientSession[F](session))
+  def apply[F[_]: Async](session: JClientSession): ClientSession[F] =
+    TransformedClientSession[F, F](session, FunctionK.id)
```
I think having at least 1/3 less code is worth it actually, in one effect function `FunctionK.id.apply` would be somewhat like 3-5 jvm commands (they could be removed btw by java itself, if they called too often), it's probably a couple of nanoseconds spent not in hot path 🤷

cryogenian commented:
+ `Bson` prefix in codecs
+ Made `getClusterDescription` effectful

wemrysi commented on core/src/main/scala/mongo4cats/client/ClientSession.scala:
```diff
@@ -16,72 +16,63 @@
-  /** Abort a transaction in the context of this session. A transaction can only be aborted if one has first been started.
-    */
+trait ClientSession[F[_]] {
+  def session: JClientSession
```
Is it necessary to expose this publicly? Maybe `private[mongo4cats]` if we need to access it internally?

cryogenian replied:
:+1:

wemrysi commented on core/src/main/scala/mongo4cats/client/MongoClient.scala:
```diff
@@ -16,55 +16,76 @@
+  def listDatabaseNames: Stream[F, String]
+  def listDatabases(
+      clientSession: Option[ClientSession[F]] = None
+  ): Stream[F, Document]
```
I think this should return `BsonDocument`, so we avoid `Document` in any of the signatures.

cryogenian replied:
Yeah, but this will be `map((x: Document) => BsonDocument.parse(x.toJson())` since `Document` is what driver returns.

cryogenian replied:
Actually, ignore my comment, it might be `listDatabase(session, classOf[BsonDocument])`

wemrysi replied:
Or even `.map(_.toBsonDocument)` since `Document <: Bson`?

wemrysi commented on core/src/main/scala/mongo4cats/collection/MongoCollection.scala:
```diff
@@ -16,572 +16,594 @@
+import cats.~>
 import cats.effect.Async
-import cats.syntax.functor._
+import cats.implicits._
```
⛳ 
```suggestion
import cats.syntax.all._
```

cryogenian replied:
:+1:

wemrysi commented on core/src/main/scala/mongo4cats/collection/queries/QueryCommand.scala:
```diff
@@ -18,209 +18,61 @@ package mongo4cats.collection.queries
 
-sealed private[queries] trait QueryCommand[O[_] <: Publisher[_], T] {
-  def run(pub: O[T]): O[T]
+sealed trait DistinctCommand extends Product with Serializable
```
Could all of these be `private[queries]`?

cryogenian replied:
:+1:

wemrysi commented on core/src/main/scala/mongo4cats/database/MongoDatabase.scala:
```diff
@@ -44,120 +40,115 @@ abstract class MongoDatabase[F[_]] {
-  def writeConcern: WriteConcern = database.getWriteConcern
-  def withWriteConcern(writeConcert: WriteConcern): MongoDatabase[F] =
-    new LiveMongoDatabase[F](database.withWriteConcern(writeConcert))
+  def listCollections(clientSession: Option[ClientSession[F]] = None): Stream[F, Document]
```
`BsonDocument` here also.

cryogenian replied:
:+1:

wemrysi commented on core/src/main/scala/mongo4cats/collection/queries/WatchQueryBuilder.scala:
```diff
@@ -16,108 +16,153 @@
+
+  //
+  def stream: Stream[F, ChangeStreamDocument[BsonValue]]
+  def boundedStream(c: Int): Stream[F, ChangeStreamDocument[BsonValue]]
```
I think we should remove `boundedStream` from all builders, it isn't well defined and I think existed in the upstream impl because they didn't have any backpressure, as such it doesn't make much sense in this impl.

cryogenian replied:
:+1:

wemrysi reviewed (commented):
A few small comments, but otherwise looks great, thanks!

rintcius commented on circe/src/main/scala/mongo4cats/circe.scala:
```diff
@@ -16,62 +16,47 @@
-          null // scalastyle:ignore
-        }
+  object unsafe {
+    def circeDocumentEncoder[A: Encoder] = new BsonDocumentEncoder[A] {
```
Why does this exist exactly? Is this for convenience only where we know usage should be ok?

cryogenian replied:
Right, this is exact reason! `BsonDocument` is required by `insert` but there is no direct mapping between it and `Json`

cryogenian commented:
Extracted `mongo4cats.circe.codecs` from `mongo4cats.circe.implicits`, it's a bit confusing that using proves that `io.circe.Encoder => mongo4cats.bson.BsonEncoder` puts custom codecs to the scope.
